cinema/views.py

```diff
@@ -46,13 +95,76 @@
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
-    serializer_class = MovieSessionSerializer
 
-    def get_serializer_class(self):
+    def get_queryset(self) -> QuerySet:
+        self.queryset = self.queryset.select_related(
+            "movie",
+            "cinema_hall"
+        ).annotate(
+            tickets_available=(
+                F("cinema_hall__rows") * F("cinema_hall__seats_in_row")
+            ) - Count("tickets")
+        )
+
+        date = self.request.query_params.get("date")
+        movie = self.request.query_params.get("movie")
+        if date:
+            try:
+                date = datetime.strptime(date, "%Y-%m-%d").date()
+            except ValueError:
+                return Response(
+                    {"error": "Invalid date format. Please use YYYY-MM-DD."},
+                    status=status.HTTP_400_BAD_REQUEST
+                )
+            self.queryset = self.queryset.filter(show_time__date=date)
+
+        if movie:
+            try:
+                movie = int(movie)
+            except ValueError:
+                return Response(
+                    {"error": "Invalid movie ID. Please provide an integer."},
+                    status=status.HTTP_400_BAD_REQUEST
+                )
```

Comment:
Similar to the previous comment, returning a `Response` object directly from `get_queryset` is not standard. Consider raising a `serializers.ValidationError` or handling the error in a different method to ensure consistency.